src/sentry/middleware/integrations/parsers/base.py

```diff
@@ -0,0 +1,110 @@
+from __future__ import annotations
+
+import abc
+import logging
+from typing import Callable, Iterable, Sequence
+
+from django.http.request import HttpRequest
+from django.urls import ResolverMatch, resolve
+
+from sentry.models.integrations import Integration, OrganizationIntegration
+from sentry.models.organization import Organization
+from sentry.silo import SiloLimit, SiloMode
+from sentry.silo.client import RegionSiloClient
+from sentry.types.region import Region, get_region_for_organization
+
+logger = logging.getLogger(__name__)
+
+
+class BaseRequestParser(abc.ABC):
+    """Base Class for Integration Request Parsers"""
+
+    def __init__(self, request: HttpRequest, response_handler: Callable):
+        self.request = request
+        self.match: ResolverMatch = resolve(self.request.path)
+        self.response_handler = response_handler
+        self.error_message = "Integration Request Parsers should only be run on the control silo."
+
+    def _ensure_control_silo(self):
+        if SiloMode.get_current_mode() != SiloMode.CONTROL:
+            logger.error(
+                "integration_control.base.silo_error",
+                extra={"path": self.request.path, "silo": SiloMode.get_current_mode()},
+            )
+            raise SiloLimit.AvailabilityError(self.error_message)
+
+    def get_response_from_control_silo(self):
+        """
+        Used to handle the request directly on the control silo.
+        """
+        self._ensure_control_silo()
+        return self.response_handler(self.request)
+
+    def get_response_from_region_silos(self, regions: Iterable[Region]):
+        """
+        Used to handle the requests on a given list of regions (synchronously).
+        If multiple regions are provided, only the last response is returned to the requestor.
+        """
+        self._ensure_control_silo()
+        region_response = None
+        for region in regions:
+            region_client = RegionSiloClient(region)
+            region_response = region_client.proxy_request(self.request).to_http_response()
+        if region_response is None:
```

Comment:
Wouldn't this only be true if the *last* region returned none?

Reply:
That makes sense, I'll change it to only change `region_response` if it isn't None.